typing/typecore.ml

```diff
@@ -7137,6 +7137,9 @@ and type_comprehension_iterator
              ~explanation:(Comprehension_in_iterator comprehension_type)
              seq_ty)
       in
+      (* TODO: fix handling of first-class module patterns so we can remove
+       * this line. *)
+      allow_modules := false;
```

Comment:
Why have this here instead of in the call to `reset_pattern` in `type_comprehension_clause`? As it stands, I have to worry about whether some patterns will have modules enabled and others won't... but then the patterns with modules enabled all have type `int`, so I guess it's OK. But much simpler if we just say in `type_comprehension_clause` "no modules here!". But perhaps I've missed an interaction that would make my approach not work...

Reply:
Calling `reset_patterns false` in [`type_comprehension_clause`](https://github.com/ncik-roberts/ocaml-jst/blob/62ad445d4e9e1fb19dbf97901292d1bdd57faacc/typing/typecore.ml#L7061) doesn't work because, just as typechecking some expressions (read: top-level modules) sets `allow_modules` to `false`, so too does typechecking other expressions (read: let-expressions) sets `allow_modules` to `true`.

In particular, the tests I added in this feature with the `let t = t in t` body would have a worse error message if we did this call in `type_comprehension_clause`.

It's true that I wasn't thinking about the int range case, but that's only because there's no corresponding call to `type_pat`. I didn't confirm this previously, but the error there looks fine:

```
# [ x for (module _ : S) = 1 to 3 ];;
Error: Invalid pattern in comprehension for-range iterator: only variables and _ are allowed.
```

Reply:
Yes, non-variable, non-underscore patterns are syntactically disallowed in `for` loops and, correspondingly, in int ranges in comprehensions

Reply:
Yes, well, that makes some sense. But it's interesting to note that `let ... and ... and ...` processes all the patterns first, then the RHSs. This allows all the patterns to have one shared `reset_pattern`, with no intervening expressions to process. In contrast, comprehensions processes the clauses one at a time, interleaving patterns with expressions -- meaning the `reset_pattern` approach doesn't work. My hunch is that other problems will arise from the difference in treatment here, in which case it will make sense to refactor this all. (It might be worth reading through `type_let` to see all the stuff that happens between pattern processing and expression processing. But probably not actually worth it -- let's just wait for bugs to crop up.)

@ncik-roberts please Resolve the conversation after reading this. No more action to take here.